2.3.0

KillerOfPie opened the pull request:
[contributing]: https://github.com/SparklingComet/TradeShop/blob/master/.github/CONTRIBUTING.md

## Pull Request Etiquette

There are several guidelines you should follow in order for your
  Pull Request to be merged.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Template

### Pull Request

Replace this sentence with general description of what your Pull Request does.

KillerOfPie committed:
Creation of and switch to using generic DataStorage object for managing data

KillerOfPie committed:
Creation of and switch to using generic DataStorage object for managing data

KillerOfPie committed:
Creation of and switch to using generic DataStorage object for managing data

KillerOfPie committed:
Changes and new data storage
 - Changes to make player data storage more versatile
 - Added relevant shops to player data file

KillerOfPie committed:
Changes and new data storage
 - Changes to make player data storage more versatile
 - Added relevant shops to player data file

KillerOfPie committed:
Changes and new data storage
 - Changes to make player data storage more versatile
 - Added relevant shops to player data file

KillerOfPie committed:
Catch for bad data setting
 - Changes to catch and stop plugin when data-storage-type is improperly set
 - Changes to Debug system to allow negative positions for error codes

KillerOfPie committed:
Catch for bad data setting
 - Changes to catch and stop plugin when data-storage-type is improperly set
 - Changes to Debug system to allow negative positions for error codes

KillerOfPie committed:
Catch for bad data setting
 - Changes to catch and stop plugin when data-storage-type is improperly set
 - Changes to Debug system to allow negative positions for error codes

KillerOfPie committed:
Added stored Chest Linkage data for Json
 - May remove persistent data values later if this works well
 - Also fixed a missing .getString for a setting in Utils

KillerOfPie committed:
Added stored Chest Linkage data for Json
 - May remove persistent data values later if this works well
 - Also fixed a missing .getString for a setting in Utils

KillerOfPie committed:
Added stored Chest Linkage data for Json
 - May remove persistent data values later if this works well
 - Also fixed a missing .getString for a setting in Utils

KillerOfPie committed:
Merge remote-tracking branch 'origin/Data-Management' into Data-Management

# Conflicts:
#	src/main/java/org/shanerx/tradeshop/utils/data/JsonConfiguration.java

KillerOfPie committed:
Merge remote-tracking branch 'origin/Data-Management' into Data-Management

# Conflicts:
#	src/main/java/org/shanerx/tradeshop/utils/data/JsonConfiguration.java

KillerOfPie committed:
Internal Permissions and shulker over trade fix
- Added support for internal permissions, disabled by default.
- Fixed bug where shops were trading all matching shulkers in their inventory instead of the proper amount.

KillerOfPie committed:
Internal Permissions and shulker over trade fix
- Added support for internal permissions, disabled by default.
- Fixed bug where shops were trading all matching shulkers in their inventory instead of the proper amount.

KillerOfPie commented:
Got rid of the data management branch and just put it all here, other way was too complicated

SparklingComet commented:
#91

KillerOfPie committed:
Added basic join message for shop statuses
* Did minor testing, seems to be working

KillerOfPie committed:
Added basic join message for shop statuses
* Did minor testing, seems to be working

KillerOfPie committed:
Additional Work on shop statuses
- Added `ts status` command
- Player gets text status on join and gui status oon command
- Changed reload command to new Manage_Plugin permission
- New methods in Message to potentially support json messages but don't know if they work for that, they do run normal messages properly, will move everything over to using them later.
- Fixed an NPE from Shop
- Fixed NPE fromShopProtectionListener when destroying iTrade signs

KillerOfPie committed:
Additional Work on shop statuses
- Added `ts status` command
- Player gets text status on join and gui status oon command
- Changed reload command to new Manage_Plugin permission
- New methods in Message to potentially support json messages but don't know if they work for that, they do run normal messages properly, will move everything over to using them later.
- Fixed an NPE from Shop
- Fixed NPE fromShopProtectionListener when destroying iTrade signs

KillerOfPie committed:
Bug Fixes + 2.2.5
- Fixed SHOP_BAD_COLOUR bug
- Fix for filled shulker duplication bug

KillerOfPie committed:
Merge branch '2.3.0' of https://github.com/SparklingComet/TradeShop into 2.3.0

 Conflicts:
	src/main/java/org/shanerx/tradeshop/utils/data/JsonConfiguration.java

KillerOfPie committed:
Added stock counter to Statuses
* updates during #saveShop so shops must be interacted with before the numbers will show

KillerOfPie committed:
Added `/ts edit` command
- Allows editing shops through an Inventory GUI
- Admins have full edit access
- Owners can change/remove users, adding still requires the command
- Owners/Managers can edit compare settings of and remove cost and products, adding/setting still requires the commands
- Other support methods and fixes

KillerOfPie committed:
Fixes
- Prevents and forcefully removes shops using Shulker Boxes as a cost when the storage inventory is a Shulker Box
- Moved Shulker Inventory checking into ShopItemStack#isSimilar
- Improved Shulker Inventory checking
- Fix for #93

KillerOfPie committed:
Fixes
- Fixes for errors during trade and destruction of iTrade shops
- Removed old shulker code

KillerOfPie committed:
`what` command separation and improvements
- clicking on items now shows the settings for it
- gui is now page based, showing 3 items per side at a time

KillerOfPie committed:
Fix for NPE with player data

KillerOfPie committed:
Minor changes

SparklingComet committed:
Fix invalid world metadata bug

SparklingComet committed:
Add PlayerSetting field to toggle involved shops motd

SparklingComet committed:
Fix missing !

SparklingComet committed:
Add missing feedback message to /ts togglestatus
and add *Barrier* to illegal-items config list

SparklingComet committed:
Add smart firework support

SparklingComet committed:
Minor change and tentative fix for old issue...
These are old changes that had not been committed unfortunately. Pushing them now.

SparklingComet committed:
Add requested setting to disable chest/sign breaking protection

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/objects/ShopItemStack.java:
```diff
@@ -227,6 +232,8 @@ public boolean isSimilar(ItemStack toCompare) {
 
 
         boolean useMeta = itemStack.hasItemMeta() == toCompare.hasItemMeta() && itemStackMeta != null, useBookMeta = itemStackBookMeta != null && toCompareBookMeta != null;
+        boolean compareFireworkDuration = Setting.FIREWORK_COMPARE_DURATION.getBoolean();
+        boolean compareFireworkEffects = Setting.FIREWORK_COMPARE_EFFECTS.getBoolean();
```
These should also be editable per shop in EditCommand.java, the last one currently in there is on line 495. I would say to remove the main setting , but if you would like to keep that then we may want to make the rest of the shop options toggle-able in the settings as well.

KillerOfPie replied:
PS. I may clean up EditCommand before you get to it so the line number may not be good anymore

SparklingComet replied:
It wasn't added in back then because I felt it would open a door for abuse as in shop owners scamming buyers by giving them short-duration or effect-less fireworks.
Then again, if buyers can view the shop settings, that would shift the responsibility towards them.

KillerOfPie committed:
Cleanup of ShopItemStack

KillerOfPie committed:
Fix for NoSuchMethodException in ShopItemStack

KillerOfPie committed:
Removed Depreciated Prefix variable and getter in Utils

KillerOfPie committed:
Message changes and removed an error suppressor

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/EditCommand.java:
```diff
@@ -281,123 +282,96 @@ private StaticGuiElement shopItemEditMenu(int index, boolean isCost) {
-                itemGroup.addElement(booleanOption("CompareEnchantments", tempItem, TRUE_ITEM, FALSE_ITEM));
-                itemGroup.addElement(booleanOption("CompareItemFlags", tempItem, TRUE_ITEM, FALSE_ITEM));
-                itemGroup.addElement(booleanOption("CompareCustomModelData", tempItem, TRUE_ITEM, FALSE_ITEM));
-                itemGroup.addElement(booleanOption("CompareAttributeModifier", tempItem, TRUE_ITEM, FALSE_ITEM));
```
what happened here exactly?

KillerOfPie replied:
I got rid of all the separate methods for this and replaced with a map<Enum key, ObjectHolder> so they are easier to retrieve and add onto later when new setting(like the fireworks) need to be added. These methods changed from their strings to the Enum values, although I didn't full caps them

KillerOfPie committed:
Minor bug fixes from Discord testing

KillerOfPie committed:
Minor bug fixes from Discord testing

KillerOfPie committed:
Minor bug fixes from Discord testing
- Fix for Random error from hoppers events with null block
- Potential Fix for Shulker Dupe, untested

KillerOfPie committed:
Fix for incorrect Max Shop Per chunk counting

KillerOfPie committed:
Fixes and Minor Changes
- Fix for illegal-items not being properly blocked in at least 1.17
- Other small changes and clean up done while searching for above
- Added additional check in add/set cost/product methods to prevent blacklist items from being added
- Change to SHOP_EMPTY Message

KillerOfPie committed:
Added amount comments to DebugLevels.java

KillerOfPie committed:
Minor Cleanups

KillerOfPie committed:
Re-Organization

Moved SettingSectionKeys to its own class so it can be used by add-on plugins

Moved IllegalWorldException to new exceptions package

KillerOfPie committed:
Update BStats and Generalized Updater for Add-on usage

KillerOfPie committed:
Fix for ShopChest#resetName NPE

SparklingComet committed:
TradeShopReloadEvent and rename the trading events

SparklingComet committed:
Merge remote-tracking branch 'origin/2.3.0' into 2.3.0

# Conflicts:
#	src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java
#	src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java

SparklingComet committed:
Resolve mess made by IDEA due to merge conflicts

SparklingComet committed:
Remove stacktrace if Updater fails and add exception msg to warning

SparklingComet committed:
Fix hopper flow shop closing issue and add FROM_HOPPER protection

KillerOfPie committed:
Ignore Hopper events if all Hopper settings are set to allow.

KillerOfPie committed:
Added CommandSender to Reload event

KillerOfPie committed:
Fix for error when Owner name is missing.

KillerOfPie committed:
Fixes and Bundle Support
- Now removes any materials from the autofill list that are not obtainable through game or commands
- Added checks for Bundle Items if version is at least 1.17

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java:
```diff
@@ -166,13 +166,18 @@ public void onBlockInteract(PlayerInteractEvent e) {
+        String owner = shop.getOwner().getName();
+
+        if (owner == null)
+            owner = "-Unknown-";
```
Is `owner` ever expected to be null or is this just an instance of defensive programming?

KillerOfPie replied:
It is never expected and should never happen, however, I ran into on Dark Chromas test server. Hence the fix.

SparklingComet replied:
Hmm, that’s odd…

KillerOfPie replied:
It could be that the servers player data was not copied over to their dev server so the players name didn't exist.

KillerOfPie committed:
Fix for Error in ts-LOG on Paper

KillerOfPie committed:
Potential fix for empty player data files

KillerOfPie committed:
Changed PlayerSuccessfulTradeEvent to use the ItemStacks that were traded instead of the ShopItemStacks that came from the shop.

KillerOfPie committed:
Fix for Name Checking working improperly/not at all.
- small Code Cleanup as well

SparklingComet committed:
change version tag from DEV to STABLE
